File: sdks/go-sdk/utils/units.go
// Quantaureum Go SDK source, version 1.0.0.
package utils

import (
	"math/big"
	"strings"

	"github.com/quantaureum/qau/sdks/go-sdk/errors"
)

// Common unit multipliers
var (
	// Wei is the smallest unit (1)
	Wei = big.NewInt(1)
	// GWei is 10^9 wei
	GWei = new(big.Int).Exp(big.NewInt(10), big.NewInt(9), nil)
	// Ether is 10^18 wei
	Ether = new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)
)

// ToWei converts an ether string value to wei.
// The decimals parameter specifies the number of decimal places (18 for ether).
func ToWei(value string, decimals int) (*big.Int, error) {
	if value == "" {
		return nil, errors.NewValidationError("value", "value cannot be empty")
	}

	// Handle negative values
	negative := false
	if strings.HasPrefix(value, "-") {
		negative = true
		value = value[1:]
	}

	// Split by decimal point
	parts := strings.Split(value, ".")
	if len(parts) > 2 {
		return nil, errors.NewValidationError("value", "invalid number format")
	}

	// Parse integer part
	intPart := parts[0]
	if intPart == "" {
		intPart = "0"
	}

	intVal, ok := new(big.Int).SetString(intPart, 10)
	if !ok {
		return nil, errors.NewValidationError("value", "invalid integer part")
	}

	// Calculate multiplier
	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(decimals)), nil)

	// Multiply integer part by multiplier
	result := new(big.Int).Mul(intVal, multiplier)

	// Handle decimal part if present
	if len(parts) == 2 {
		decPart := parts[1]
		if len(decPart) > decimals {
			// Truncate to decimals precision
			decPart = decPart[:decimals]
		}

		// Pad with zeros if needed
		for len(decPart) < decimals {
			decPart += "0"
		}

		decVal, ok := new(big.Int).SetString(decPart, 10)
		if !ok {
			return nil, errors.NewValidationError("value", "invalid decimal part")
		}

		result.Add(result, decVal)
	}

	if negative {
		result.Neg(result)
	}

	return result, nil
}

// FromWei converts wei to an ether string value.
// The decimals parameter specifies the number of decimal places (18 for ether).
func FromWei(wei *big.Int, decimals int) string {
	if wei == nil {
		return "0"
	}

	// Handle negative values
	negative := wei.Sign() < 0
	weiAbs := new(big.Int).Abs(wei)

	// Calculate divisor
	divisor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(decimals)), nil)

	// Get integer and remainder
	intPart := new(big.Int).Div(weiAbs, divisor)
	remainder := new(big.Int).Mod(weiAbs, divisor)

	// Format result
	result := intPart.String()

	if remainder.Sign() > 0 {
		// Format decimal part with leading zeros
		decStr := remainder.String()
		for len(decStr) < decimals {
			decStr = "0" + decStr
		}
		// Trim trailing zeros
		decStr = strings.TrimRight(decStr, "0")
		if decStr != "" {
			result += "." + decStr
		}
	}

	if negative {
		result = "-" + result
	}

	return result
}

// ParseUnits is an alias for ToWei for compatibility.
func ParseUnits(value string, decimals int) (*big.Int, error) {
	return ToWei(value, decimals)
}

// FormatUnits is an alias for FromWei for compatibility.
func FormatUnits(wei *big.Int, decimals int) string {
	return FromWei(wei, decimals)
}

// EtherToWei converts ether string to wei (18 decimals).
func EtherToWei(ether string) (*big.Int, error) {
	return ToWei(ether, 18)
}

// WeiToEther converts wei to ether string (18 decimals).
func WeiToEther(wei *big.Int) string {
	return FromWei(wei, 18)
}

// GWeiToWei converts gwei string to wei (9 decimals).
func GWeiToWei(gwei string) (*big.Int, error) {
	return ToWei(gwei, 9)
}

// WeiToGWei converts wei to gwei string (9 decimals).
func WeiToGWei(wei *big.Int) string {
	return FromWei(wei, 9)
}

File: sdks/rust-sdk/src/utils/units.rs
// Quantaureum Rust SDK source, version 1.0.0.
//! Wei/Ether unit conversion utilities.
//!
//! This module provides functions for converting between different units of cryptocurrency.
//! The base unit is Wei (1 Ether = 10^18 Wei).

use crate::error::{Error, Result};
use primitive_types::U256;

/// Number of decimal places for Ether (18 decimals).
pub const ETHER_DECIMALS: u32 = 18;

/// Converts an ether value string to Wei.
///
/// # Arguments
///
/// * `ether` - The ether value as a string (e.g., "1.5")
/// * `decimals` - The number of decimal places (typically 18 for Ether)
///
/// # Returns
///
/// A `Result` containing the Wei value as U256.
///
/// # Errors
///
/// Returns an error if the input string is invalid or the value overflows.
///
/// # Examples
///
/// ```
/// use quantaureum_sdk::utils::to_wei;
///
/// let wei = to_wei("1.0", 18).unwrap();
/// assert_eq!(wei.to_string(), "1000000000000000000");
///
/// let wei = to_wei("0.5", 18).unwrap();
/// assert_eq!(wei.to_string(), "500000000000000000");
/// ```
pub fn to_wei(ether: &str, decimals: u32) -> Result<U256> {
    parse_units(ether, decimals)
}

/// Converts a Wei value to an ether string.
///
/// # Arguments
///
/// * `wei` - The Wei value as U256
/// * `decimals` - The number of decimal places (typically 18 for Ether)
///
/// # Returns
///
/// The ether value as a string.
///
/// # Examples
///
/// ```
/// use quantaureum_sdk::utils::from_wei;
/// use primitive_types::U256;
///
/// let ether = from_wei(U256::from(1_000_000_000_000_000_000u64), 18);
/// assert_eq!(ether, "1.0");
///
/// let ether = from_wei(U256::from(500_000_000_000_000_000u64), 18);
/// assert_eq!(ether, "0.5");
/// ```
pub fn from_wei(wei: U256, decimals: u32) -> String {
    format_units(wei, decimals)
}

/// Parses a decimal string value into the smallest unit.
///
/// # Arguments
///
/// * `value` - The decimal value as a string (e.g., "1.5")
/// * `decimals` - The number of decimal places
///
/// # Returns
///
/// A `Result` containing the value in the smallest unit as U256.
///
/// # Errors
///
/// Returns an error if:
/// - The input string is empty or invalid
/// - The value has more decimal places than allowed
/// - The value overflows U256
pub fn parse_units(value: &str, decimals: u32) -> Result<U256> {
    // Validate decimal string format
    crate::utils::validation::validate_decimal_string(value, "value")?;

    let value = value.trim();

    // Split into integer and fractional parts
    let parts: Vec<&str> = value.split('.').collect();

    // Already validated that there's at most one decimal point

    let integer_part = parts[0];
    let fractional_part = if parts.len() == 2 { parts[1] } else { "" };

    // Check if fractional part is too long
    if fractional_part.len() > decimals as usize {
        return Err(Error::Validation {
            field: "value".to_string(),
            message: format!(
                "too many decimal places: {} (max {})",
                fractional_part.len(),
                decimals
            ),
        });
    }

    // Calculate the multiplier
    let multiplier = U256::from(10).pow(U256::from(decimals));

    // Parse integer part
    let integer_value = if integer_part.is_empty() || integer_part == "-" {
        U256::zero()
    } else {
        U256::from_dec_str(integer_part.trim_start_matches('-')).map_err(|_| Error::Validation {
            field: "value".to_string(),
            message: "integer part overflow".to_string(),
        })?
    };

    // Calculate integer contribution
    let integer_wei = integer_value
        .checked_mul(multiplier)
        .ok_or_else(|| Error::Validation {
            field: "value".to_string(),
            message: "overflow".to_string(),
        })?;

    // Parse fractional part
    let fractional_wei = if fractional_part.is_empty() {
        U256::zero()
    } else {
        // Pad fractional part to full decimals
        let padded = format!("{:0<width$}", fractional_part, width = decimals as usize);
        U256::from_dec_str(&padded).map_err(|_| Error::Validation {
            field: "value".to_string(),
            message: "fractional part overflow".to_string(),
        })?
    };

    // Combine
    let result = integer_wei
        .checked_add(fractional_wei)
        .ok_or_else(|| Error::Validation {
            field: "value".to_string(),
            message: "overflow".to_string(),
        })?;

    Ok(result)
}

/// Formats a value from the smallest unit to a decimal string.
///
/// # Arguments
///
/// * `value` - The value in the smallest unit as U256
/// * `decimals` - The number of decimal places
///
/// # Returns
///
/// The formatted decimal string.
pub fn format_units(value: U256, decimals: u32) -> String {
    if decimals == 0 {
        return value.to_string();
    }

    let divisor = U256::from(10).pow(U256::from(decimals));
    let integer_part = value / divisor;
    let fractional_part = value % divisor;

    if fractional_part.is_zero() {
        format!("{}.0", integer_part)
    } else {
        // Format fractional part with leading zeros
        let frac_str = format!("{:0>width$}", fractional_part, width = decimals as usize);
        // Trim trailing zeros
        let frac_str = frac_str.trim_end_matches('0');
        format!("{}.{}", integer_part, frac_str)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    #[test]
    fn test_to_wei_whole_number() {
        let wei = to_wei("1", 18).unwrap();
        assert_eq!(wei, U256::from(10u64).pow(U256::from(18)));
    }

    #[test]
    fn test_to_wei_decimal() {
        let wei = to_wei("1.5", 18).unwrap();
        let expected = U256::from(15u64) * U256::from(10u64).pow(U256::from(17));
        assert_eq!(wei, expected);
    }

    #[test]
    fn test_to_wei_small_decimal() {
        let wei = to_wei("0.000000000000000001", 18).unwrap();
        assert_eq!(wei, U256::from(1));
    }

    #[test]
    fn test_to_wei_zero() {
        let wei = to_wei("0", 18).unwrap();
        assert_eq!(wei, U256::zero());
    }

    #[test]
    fn test_from_wei_whole_number() {
        let ether = from_wei(U256::from(10u64).pow(U256::from(18)), 18);
        assert_eq!(ether, "1.0");
    }

    #[test]
    fn test_from_wei_decimal() {
        let wei = U256::from(15u64) * U256::from(10u64).pow(U256::from(17));
        let ether = from_wei(wei, 18);
        assert_eq!(ether, "1.5");
    }

    #[test]
    fn test_from_wei_small() {
        let ether = from_wei(U256::from(1), 18);
        assert_eq!(ether, "0.000000000000000001");
    }

    #[test]
    fn test_from_wei_zero() {
        let ether = from_wei(U256::zero(), 18);
        assert_eq!(ether, "0.0");
    }

    #[test]
    fn test_parse_units_different_decimals() {
        // 6 decimals (like USDC)
        let value = parse_units("1.5", 6).unwrap();
        assert_eq!(value, U256::from(1_500_000u64));
    }

    #[test]
    fn test_format_units_different_decimals() {
        // 6 decimals (like USDC)
        let value = format_units(U256::from(1_500_000u64), 6);
        assert_eq!(value, "1.5");
    }

    #[test]
    fn test_to_wei_invalid_empty() {
        assert!(to_wei("", 18).is_err());
    }

    #[test]
    fn test_to_wei_invalid_format() {
        assert!(to_wei("1.2.3", 18).is_err());
        assert!(to_wei("abc", 18).is_err());
    }

    #[test]
    fn test_to_wei_too_many_decimals() {
        assert!(to_wei("1.0000000000000000001", 18).is_err());
    }

    proptest! {
        #![proptest_config(ProptestConfig::with_cases(100))]

        // Feature: quantaureum-rust-sdk, Property 5: Wei/Ether Conversion Round-Trip
        // For any valid Wei value, calling from_wei() then to_wei() SHALL return the original value.
        // Validates: Requirements 5.1, 5.2
        #[test]
        fn prop_wei_ether_roundtrip(wei_value in 0u64..1_000_000_000_000_000_000u64) {
            let wei = U256::from(wei_value);
            let ether_str = from_wei(wei, 18);
            let recovered_wei = to_wei(&ether_str, 18).unwrap();
            prop_assert_eq!(wei, recovered_wei);
        }

        // Feature: quantaureum-rust-sdk, Property 5: Wei/Ether Conversion Round-Trip
        // For different decimal values (6, 8, 18), round-trip conversion should preserve value.
        // Validates: Requirements 5.1, 5.2
        #[test]
        fn prop_units_roundtrip_various_decimals(
            value in 0u64..1_000_000_000u64,
            decimals in 1u32..18u32
        ) {
            let wei = U256::from(value);
            let formatted = format_units(wei, decimals);
            let recovered = parse_units(&formatted, decimals).unwrap();
            prop_assert_eq!(wei, recovered);
        }
    }
}
